### PR TITLE
Collection of patches to do with --working-dir 

### PR DESCRIPTION
**Pull request by mpickering**
See individual commits for details

### CODE REVIEW

**Review by sheaf — COMMENTED** [review]
Could you audit all uses of `runDbProgram` and change them to `runDbProgramCwd` when appropriate? For instance I am seeing one usage in `Distribution.Simple.GHCJS.installExe` for instance (sigh).

**Comment by geekosaur** [comment]
It would need to be completely rewritten to be backported, as it's heavily dependent on #9718.

**Comment by mpickering** [comment]
The patch doesn't fix any issues present in 3.12.

File: Cabal/src/Distribution/Simple/HaskellSuite.hs [hunk 1]
```diff
@@ -184,8 +186,8 @@ buildLib verbosity pkg_descr lbi lib clbi = do
       language = fromMaybe Haskell98 (defaultLanguage bi)
       progdb = withPrograms lbi
       pkgid = packageId pkg_descr
-      i = interpretSymbolicPathLBI lbi -- See Note [Symbolic paths] in Distribution.Utils.Path
-  runDbProgram verbosity haskellSuiteProgram progdb $
+      i = interpretSymbolicPathCWD -- See Note [Symbolic paths] in Distribution.Utils.Path
```
Comment (sheaf) [inline Cabal/src/Distribution/Simple/HaskellSuite.hs:189]: Could you rename `i` to `u`? That's the convention I've been using when we are using `interpretSymbolicPathCWD`.

File: Cabal/src/Distribution/Simple/UHC.hs [hunk 2]
```diff
@@ -245,7 +251,7 @@ buildExe verbosity _pkg_descr lbi exe clbi = do
   userPkgDir <- getUserPackageDir
   let mbWorkDir = mbWorkDirLBI lbi
   srcMainPath <- findFileCwd verbosity mbWorkDir (hsSourceDirs $ buildInfo exe) (modulePath exe)
-  let runUhcProg = runDbProgram verbosity uhcProgram (withPrograms lbi)
+  let runUhcProg = runDbProgramCwd verbosity (mbWorkDirLBI lbi) uhcProgram (withPrograms lbi)
       i = interpretSymbolicPathLBI lbi -- See Note [Symbolic paths] in Distribution.Utils.Path
```
Comment (sheaf) [inline Cabal/src/Distribution/Simple/UHC.hs:255]: This should probably be `u = interpretSymbolicPathCWD`.

File: Cabal/src/Distribution/Simple/UHC.hs [hunk 3]
```diff
@@ -306,13 +312,13 @@ constructUHCCmdLine user system lbi bi clbi odir verbosity =
           MaximumOptimisation -> ["-O2"]
        )
   where
-    i = interpretSymbolicPathLBI lbi -- See Note [Symbolic paths] in Distribution.Utils.Path
+    i = interpretSymbolicPathCWD -- See Note [Symbolic paths] in Distribution.Utils.Path
```
Comment (sheaf) [inline Cabal/src/Distribution/Simple/UHC.hs:315]: Please rename to `u = interpretSymbolicPathCWD`.

File: Cabal/src/Distribution/Simple/HaskellSuite.hs (hunk 1 above)
Comment (mpickering) [reply Cabal/src/Distribution/Simple/HaskellSuite.hs:189]: done.

**Review by sheaf — APPROVED** [review]

File: cabal-install/src/Distribution/Client/ProjectPlanning.hs [hunk 4]
```diff
@@ -3909,22 +3909,31 @@ computeInstallDirs storeDirLayout defaultInstallDirs elaboratedShared elab
...
   (ReadyPackage elab@ElaboratedConfiguredPackage{..})
   sharedConfig@ElaboratedSharedConfig{..}
-  configCommonFlags =
-    sanityCheckElaboratedConfiguredPackage
-      sharedConfig
-      elab
-      Cabal.ConfigFlags{..}
+  configCommonFlags = do
+    -- explicitly clear, then our package db stack
+    -- TODO: [required eventually] have to do this differently for older Cabal versions
```
Comment (geekosaur) [inline cabal-install/src/Distribution/Client/ProjectPlanning.hs:3930]: How much trouble is this going to cause now, though?

**Review by geekosaur — APPROVED** [review]

**Comment by ulysses4ever** [comment]
On today's Cabal meeting, we discussed what could let this [API-changing](https://github.com/haskell/cabal/issues/10559) PR stay under the radar of 3.14.0.0's release notes. Unanimously people cited the absence of the PR template. We'll try to be more diligent about that. But we also ask the contributors to keep the template (and in particular choose one of the two templates) when submitting PRs.

**Comment by geekosaur** [comment]
There are ways to cut PRs that don't take the template into account, notably the Git integration in VSCode/VSCodium. I think I've encountered a few others but don't recall offhand what they were, just that they involved third party tooling instead of the GitHub website.

**Comment by ulysses4ever** [comment]
@geekosaur that's a great point, I didn't think about it. I myself have dabbled with Magit Forge, which is an Emacs extension allowing, among other things, [to create PRs](https://magit.vc/manual/forge/Creating-Topics-and-Posts.html). 

Oh well... I guess, the reviewers will have to pay more attention if we want a better future.

**Pull request merged**